core/templates/filters/translate.pipe.ts

```diff
@@ -0,0 +1,133 @@
+ * The object following the pipe, i.e.{x: 'val'}, is another argument to the
+ * transform function. This object is called params or interpolationParams.
+ */
+
+@Pipe({
+  name: 'translate',
+  // This option is required to update the value when promises are resolved.
+  pure: false
+})
+export class TranslatePipe implements PipeTransform, OnDestroy {
+  interpolatedValue: string = '';
+  lastKey: string;
+  lastParams: Object | Array<Object>;
+  onLangChange: Subscription;
+
+  constructor(
+    private translate: TranslateService,
+    private _ref: ChangeDetectorRef,
+    private utils: UtilsService
+  ) {}
+
+  /**
+   * @param {string} key - key for i18n translation
+   * @param {Object} interpolateParams - Params for interpolation
+   */
+  updateInterpolatedValue(key: string, interpolateParams?: Object): void {
+    const interpolatedString = this.translate.getInterpolatedString(
+      key, interpolateParams);
+
+    // Using multiline ternary (https://eslint.org/docs/rules/multiline-ternary).
+    this.interpolatedValue = interpolatedString !== undefined ?
+    interpolatedString :
+    key;
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+    this._ref.markForCheck();
+  }
+
+  /**
+   * @param {string} key - key for i18n
+   * @param {Object} params - params for interpolation
+   * @returns {string} - Interpolated I18n value
+   */
+  transform(key: string, params?: Object): string {
+    if (!key || !key.length) {
+      return key;
+    }
+
+    // If the key and params are same, return the last stored value.
+    if (key === this.lastKey &&
+          this.utils.isEquivalent(params, this.lastParams)) {
+      return this.interpolatedValue;
+    }
+
+    let interpolateParams: Object;
+    if (this.utils.isDefined(params)) {
+      if (typeof params === 'object' && !Array.isArray(params)) {
+        interpolateParams = params;
+      }
+    }
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+
+    // Storing the params to check if the params are same when the transform is
+    // invoked again.
+    this.lastParams = params;
+
+    // Update the interpolated value.
+    this.updateInterpolatedValue(key, interpolateParams);
+
+    // If there is a subscription to onLangChange, unsubscribe.
+    this._dispose();
+    // Subscribe to onLangChange event, in case the language changes.
+    if (!this.onLangChange) {
+      this.onLangChange = this.translate.onLangChange.subscribe(
+        (event: LangChangeEvent) => {
+          this.updateInterpolatedValue(key, interpolateParams);
+        }
+      );
+    }
+    return this.interpolatedValue;
+  }
+
+  /**
+   * Clean up any existing subscription to change events
+   */
+  private _dispose(): void {
+    if (typeof this.onLangChange === 'undefined') {
+      return;
+    }
```

Comment:
Hmm, I didn't really mean this, I meant something like `if (typeof this.onLangChange === 'subscription)' (I'm not sure what is the exact type here) then do the stuff that is needed.